keps/0007-kic-performance-test.md

```diff
@@ -0,0 +1,557 @@
+---
+title: KIC Performance Test Solution
+status: implementable
+---
+
+# Notes
+For references see the dependencies related to this proprosal to check the progress of related efforts:
+- solution 
+- deployment and configuration
+- code/repo
+- configuration and execution
+- profiling (code profiler)
+- histogram execution time and calculates percentiles
+
+
+## Summary
+  We want a performance test solution that can collect controller metrics per release, which also provides profiling
+and histogram execution time. The solution is able to scale according to controller scalability. We mainly use kind as test cluster, but also support varied cloud provider cluster such EKS, AKS, GKE etc.
+
+[kic]:https://github.com/kong/kubernetes-ingress-controller
+[ktf]:https://github.com/Kong/kubernetes-testing-framework
+[jenkins]: https://www.jenkins.io/
+[aws]: https://aws.amazon.com/
+[kind]: https://kind.sigs.k8s.io/
+[docker]: https://www.docker.com/ 
+
+## Motivation
+  - provide an isolated and reproducible environment for performance testing
+  - continue integration and continue test
+  - Visualize testing configuration and results.
+  - System distribution
+
+### Goals
+  - Able to run performance pipeline repeatedly
+  - Able to collect following performance metrics
+    - controller profiling
+    - controller memory/cpu consumption
+    - ingress config, deck sync, parser cost
```

Comment:
I think we should point out "what" is "able to do X" so that our goals are clearly providing some direction towards the eventual implementation details. I also personally would like to point out exactly how we intend to implement this at a very high level:

```suggestion
- Develop [Golang](https://go.dev) libraries that are able to run a performance testing pipeline on a Kubernetes cluster and collect the following metrics:
  - controller memory/cpu consumption
  - ingress config, deck sync, parser cost
  - controller profiling (pprof data)
- Develop a CLI interface to the Golang libraries for general purpose usage
- Develop a container image to run the CLI in a container for automated deployments
- Develop a [Helm Chart](https://helm.sh) for the container image to easily deploy it to Kubernetes clusters
```

Reply:
Would say the Performance Test Solution is mainly for performance testing purpose, docker container, pipeline fits well, helm chart is good suggestion though needs think about it. 
Interms of the CLM, pipeline works as similar, and the configuration items are similar as command line parameters. Would take that as an enhancement.